books/kestrel/c/syntax/parser.lisp

```diff
@@ -4712,6 +4713,136 @@
+          ((erp char pos parstate) (read-char parstate)))
+       (cond
+        ((not char) ; EOF
+         (reterr-msg :where (position-to-msg pos)
+                     :expected "a character"
+                     :found (char-to-msg char)
+                     ;; TODO GJ: update
+                     :extra (msg "The preprocessing directive starting at ~@1 ~
+                                  never ends."
+                                 (position-to-msg first-pos))))
+        ((= char 10) ; new-line
+         (retok pos parstate))
+        (t ; other
+         (lex-prepr-directive-loop first-pos parstate))))
+     :measure (parsize parstate)
+     :hints (("Goal" :in-theory (enable o< o-finp)))
+     :guard-hints (("Goal" :in-theory (enable acl2-numberp-when-natp)))
+
+     ///
+
+     (defret parsize-of-lex-prepr-directive-loop-uncond
+       (<= (parsize new-parstate)
+           (parsize parstate))
+       :rule-classes :linear
+       :hints (("Goal" :induct t)))
+
+     (defret parsize-of-lex-prepr-directive-loop-cond
+       (implies (not erp)
+                (<= (parsize new-parstate)
+                    (1- (parsize parstate))))
+       :rule-classes :linear
+       :hints (("Goal" :induct t)))))
+
+  ///
+
+  (defret parsize-of-lex-prepr-directive-uncond
+    (<= (parsize new-parstate)
+        (parsize parstate))
+    :rule-classes :linear)
+
+  (defret parsize-of-lex-prepr-directive-cond
+    (implies (not erp)
+             (<= (parsize new-parstate)
+                 (1- (parsize parstate))))
+    :rule-classes :linear))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define only-whitespace-backward-through-line ((parstate parstatep))
+  :returns (only-whitespace booleanp)
+  :short "Check that the only preceding characters on the line are whitespace."
+  :long
+  (xdoc::topstring
+   (xdoc::p
+    "We begin with the character immediately before the last read character,
+     and check that every character is whitespace until we reach either
+     a new-line or the start of the file."))
+  (b* ((chars-read (parstate->chars-read parstate))
+       ((when (= chars-read 1))
+        t)
+       ((when (or (< chars-read 1)
+                  (<= (parstate->chars-length parstate) chars-read)))
+        (raise "Internal error: chars-read index ~x0 out of bound ~x1."
+               chars-read
+               (parstate->chars-length parstate))))
+    (only-whitespace-backward-through-line-loop (- chars-read 2) parstate))
+  :guard-hints (("Goal" :in-theory (enable natp)))
+
+  :prepwork
+
+  ((define only-whitespace-backward-through-line-loop ((i natp)
+                                                       (parstate parstatep))
+     :parents nil
+     :guard (< i (parstate->chars-length parstate))
+     :returns (all-whitespace booleanp)
+     (b* ((char+pos (parstate->char i parstate))
+          (char (char+position->char char+pos)))
+       ;; TODO: need to support other newlines?
```

Comment:
We only recognize CR and/or LF as new-line currently, so I'd say that we don't need to worry about this until/if we see some code with some other kind of new-line (Unicode).